src/main/java/com/intuit/graphql/orchestrator/batch/DownstreamQueryModifier.java

```diff
@@ -0,0 +1,182 @@
+        context.setVar(GraphQLType.class, fieldDefinition.getType());
+      }
+      return TraversalControl.CONTINUE;
+    }
+  }
+
+  private boolean isExternalField(String parentTypename, String fieldName) {
+    FieldCoordinates fieldCoordinates = coordinates(parentTypename, fieldName);
+    return serviceMetadata.isOwnedByEntityExtension(fieldCoordinates);
+  }
+
+  private GraphQLFieldDefinition getFieldDefinition(
+      String name, GraphQLFieldsContainer parentType) {
+    if (TypeNameMetaFieldDef.getName().equals(name)) {
+      return TypeNameMetaFieldDef;
+    }
+    return parentType.getFieldDefinition(name);
+  }
+
+  @Override
+  public TraversalControl visitFragmentDefinition(
+      FragmentDefinition node, TraverserContext<Node> context) {
+    // if modifying selection set in a fragment definition, this will be the first code to visit.
+    context.setVar(GraphQLType.class, rootType);
+    return TraversalControl.CONTINUE;
+  }
+
+  @Override
+  public TraversalControl visitInlineFragment(InlineFragment node, TraverserContext<Node> context) {
+    GraphQLType parentType = getParentType(context);
+    context.setVar(GraphQLType.class, parentType);
+    return TraversalControl.CONTINUE;
+  }
+
+  @Override
+  public TraversalControl visitSelectionSet(SelectionSet node, TraverserContext<Node> context) {
+    GraphQLFieldsContainer parentType = (GraphQLFieldsContainer) getParentType(context);
+    context.setVar(GraphQLType.class, parentType);
+    String parentTypeName = parentType.getName();
+
+    Set<Field> selectedFields = this.selectionCollector.collectFields(node);
+
+    RequiredFieldsCollector fedRequiredFieldsCollector = RequiredFieldsCollector
+        .builder()
+        .parentTypeName(parentTypeName)
+        .serviceMetadata(this.serviceMetadata)
+        .fieldsWithResolver(getFieldsWithResolverDirective(selectedFields, parentType))
+        .fieldsWithRequiresDirective(getFieldsWithRequiresDirective(selectedFields, parentTypeName))
+        .build();
+
+    Set<Field> fieldsToAdd = fedRequiredFieldsCollector.get().stream()
+        .filter(field -> !selectedFields.contains(field))
+        .collect(Collectors.toSet());
+
+    if (CollectionUtils.isNotEmpty(fieldsToAdd)) {
+      SelectionSet newNode = node.transform(builder -> {
+        for (Field field : fieldsToAdd) {
+          builder.selection(field);
+        }
+      });
+      return changeNode(context, newNode);
+    }
+
+    return TraversalControl.CONTINUE;
+  }
+
+  private Set<GraphQLFieldDefinition> getFieldsWithResolverDirective(Set<Field> selections, GraphQLFieldsContainer parentType) {
+    if (CollectionUtils.isEmpty(selections)) {
+      return Collections.emptySet();
+    }
+
+    return selections.stream()
+        .filter(field -> !IntrospectionUtil.INTROSPECTION_FIELDS.contains(field.getName()))
+        .map(field -> parentType.getFieldDefinition(field.getName()))
+        .filter(FieldResolverDirectiveUtil::hasResolverDirective)
```

Comment:
can we calculate this from the metadata?

Reply:
I guess no.  The selections are runtime based on from selection set.  In here, we're looking ahead which selected fields has @resolver.